sds/src/secondary_validation/jwt_claims_checker.rs

```diff
@@ -0,0 +1,233 @@
+use crate::secondary_validation::Validator;
+use crate::scanner::regex_rule::config::{ClaimRequirement, JwtClaimsCheckerConfig};
+use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
+use regex::Regex;
+use serde_json::Value as JsonValue;
+use ahash::AHashMap;
+
+pub struct JwtClaimsChecker {
+    pub required_claims: Vec<(String, ClaimRequirement)>,
+    patterns: AHashMap<String, Regex>,
+}
+
+impl JwtClaimsChecker {
+    pub fn new(config: JwtClaimsCheckerConfig) -> Self {
+        let mut patterns = AHashMap::new();
+        for (claim_name, requirement) in &config.required_claims {
+            if let ClaimRequirement::RegexMatch(pattern) = requirement {
+                patterns.insert(claim_name.clone(), Regex::new(pattern).unwrap());
+            }
+        }
+        Self { required_claims: config.required_claims.into_iter().collect(), patterns }
+    }
+}
+
+impl Validator for JwtClaimsChecker {
+    fn is_valid_match(&self, regex_match: &str) -> bool {
+        if let Some((_, payload)) = decode_segments(regex_match) {
+            validate_required_claims(&payload, &self.required_claims, &self.patterns)
+        } else {
+            // If JWT segments cannot be decoded, the JWT is not well formatted
+            false
+        }
+    }
+}
+
+// This function is an extraction of the decoding part from jwt_expiration_checker
+// Reusing the JWT decoding logic to avoid duplication
+fn decode_segments(encoded_token: &str) -> Option<(JsonValue, JsonValue)> {
+    let mut raw_segments: std::str::Split<&str> = encoded_token.split(".");
+
+    let header_segment = raw_segments.next()?;
+    let payload_segment = raw_segments.next()?;
+
+    if raw_segments.next().is_none() {
+        // Invalid JWT header + payload + signature
+        // We're not explicitly checking the validity of the signature
+        // We could embed signature verification at some point if necessary
+        return None;
+    }
+
+    let header_json = decode_segment(header_segment)?;
+    let payload_json = decode_segment(payload_segment)?;
+    Some((header_json, payload_json))
+}
+
+fn decode_segment(segment: &str) -> Option<JsonValue> {
+    let decoded = URL_SAFE_NO_PAD
+        .decode(segment)
+        .ok()?;
+    serde_json::from_slice(&decoded).ok()
+}
+
+fn validate_required_claims(payload: &JsonValue, required_claims: &Vec<(String, ClaimRequirement)>, patterns: &AHashMap<String, Regex>) -> bool {
+    if let Some(payload_obj) = payload.as_object() {
+        // Check each required claim
+        required_claims.iter().all(|(claim_name, requirement)| {
+            if let Some(claim_value) = payload_obj.get(claim_name) {
+                validate_claim_requirement(claim_value, requirement, patterns.get(claim_name))
```

Comment:
nit: It's probably a bit easier to pass in a reference to the entire `patterns` map, so `validate_claim_requirement` can check for the regex pattern only if it's needed. Right now it fetches it for every claim.